File: sihcp-server/src/main/java/com/inventario/sihcp/mobile/server/dto/PagedResponseDTO.java
package com.inventario.sihcp.mobile.server.dto;

import java.util.List;

/**
 * DTO de resposta paginada dedicado ao endpoint
 * {@code GET /api/mobile/descricoes/sugestoes}.
 *
 * <p>Diferencia-se de {@link PagedResponse} por ser um {@code record} Java 21
 * com um contrato reduzido e explícito ao app mobile, incluindo o indicador
 * {@link #semInventarioAtivo()} usado quando não há inventário ativo no
 * momento da requisição (Req 5.8).</p>
 *
 * <p>Campos de metadados (Req 5.5, 5.6):</p>
 * <ul>
 *   <li>{@code content}        — itens da página atual (nunca {@code null})</li>
 *   <li>{@code page}           — número da página solicitada (0-based)</li>
 *   <li>{@code size}           — tamanho efetivo aplicado (após sanitização)</li>
 *   <li>{@code totalElements}  — total de elementos em todas as páginas</li>
 *   <li>{@code totalPages}     — {@code ceil(totalElements / size)}</li>
 *   <li>{@code hasNext}        — {@code (page + 1) * size < totalElements}</li>
 *   <li>{@code semInventarioAtivo} — {@code true} quando não existe inventário
 *       ativo; nesse caso {@code content} é vazio e os demais metadados são
 *       zerados (Req 5.8)</li>
 * </ul>
 *
 * <p>Requirements: 5.5, 5.6, 5.8.</p>
 *
 * @param <T> tipo dos elementos em {@code content}
 * @param content             itens da página atual
 * @param page                número da página (0-based)
 * @param size                tamanho efetivo da página
 * @param totalElements       total de elementos
 * @param totalPages          total de páginas
 * @param hasNext             indica se há próxima página
 * @param semInventarioAtivo  indica ausência de inventário ativo (Req 5.8)
 */
public record PagedResponseDTO<T>(
        List<T> content,
        int page,
        int size,
        long totalElements,
        int totalPages,
        boolean hasNext,
        boolean semInventarioAtivo
) {

    /**
     * Fábrica para o caso {@code semInventarioAtivo=true} (Req 5.8).
     *
     * <p>Retorna uma resposta com {@code content} vazio e todos os metadados
     * de paginação zerados. O flag {@link #semInventarioAtivo()} permite ao
     * app diferenciar "nenhum resultado encontrado" de "não há inventário
     * ativo para consultar".</p>
     *
     * @param <T> tipo dos elementos
     * @return resposta vazia com {@code semInventarioAtivo=true}
     */
    public static <T> PagedResponseDTO<T> empty() {
        return new PagedResponseDTO<>(List.of(), 0, 0, 0L, 0, false, true);
    }

    /**
     * Variante explícita da fábrica vazia.
     *
     * <p>Utilidade principal: permitir a construção de respostas vazias
     * tanto no cenário "sem inventário ativo" ({@code semInventarioAtivo=true},
     * Req 5.8) quanto em cenários em que simplesmente não há resultados
     * com inventário ativo presente ({@code semInventarioAtivo=false}).</p>
     *
     * @param semInventarioAtivo valor do flag de ausência de inventário ativo
     * @param <T> tipo dos elementos
     * @return resposta vazia com o flag informado
     */
    public static <T> PagedResponseDTO<T> empty(boolean semInventarioAtivo) {
        return new PagedResponseDTO<>(List.of(), 0, 0, 0L, 0, false, semInventarioAtivo);
    }
}

File: InventarioMobile/app/src/main/java/com/inventario/mobile/data/remote/dto/PagedResponseDto.kt
package com.inventario.mobile.data.remote.dto

import com.google.gson.annotations.SerializedName

/**
 * DTO genérico de resposta paginada dedicado ao endpoint
 * `GET /api/mobile/descricoes/sugestoes` da feature
 * "coleta com descrição livre e sugestão".
 *
 * Espelha, campo a campo, o record `PagedResponseDTO<T>` definido no
 * servidor (`sihcp-server/.../mobile/server/dto/PagedResponseDTO.java`).
 * É um tipo **distinto** do existente [PagedResponse] (que continua sendo
 * usado pela implementação Paging 3 de patrimônios/coletas) porque:
 *
 *   1. contém o campo adicional `semInventarioAtivo` (Req 5.8),
 *   2. omite os campos `first`, `last` e `hasPrevious` do contrato legado,
 *   3. o contrato JSON do novo endpoint é diferente do antigo.
 *
 * Campos (Req 5.5, 5.6, 5.8):
 *   - `content`             itens da página atual
 *   - `page`                número da página (0-based)
 *   - `size`                tamanho efetivo da página (após sanitização)
 *   - `totalElements`       total de elementos em todas as páginas
 *   - `totalPages`          ceil(totalElements / size)
 *   - `hasNext`             (page + 1) * size < totalElements
 *   - `semInventarioAtivo`  true quando não há inventário ativo; nesse
 *                           caso `content` é vazio e os demais metadados
 *                           são zerados (Req 5.8)
 *
 * Tipos são plain Kotlin / boxed nulos onde faz sentido, para máxima
 * tolerância na desserialização via Gson/Retrofit.
 *
 * Requirements: 5.5, 5.6, 5.8.
 */
data class PagedResponseDto<T>(
    @SerializedName("content")
    val content: List<T> = emptyList(),

    @SerializedName("page")
    val page: Int = 0,

    @SerializedName("size")
    val size: Int = 0,

    @SerializedName("totalElements")
    val totalElements: Long = 0L,

    @SerializedName("totalPages")
    val totalPages: Int = 0,

    @SerializedName("hasNext")
    val hasNext: Boolean = false,

    @SerializedName("semInventarioAtivo")
    val semInventarioAtivo: Boolean = false
) {
    companion object {
        /**
         * Fábrica para o caso "sem inventário ativo" (Req 5.8).
         *
         * Retorna uma resposta com `content` vazio, metadados zerados e
         * `semInventarioAtivo=true`. Espelha `PagedResponseDTO.empty()`
         * do servidor.
         */
        fun <T> empty(): PagedResponseDto<T> = PagedResponseDto(
            content = emptyList(),
            page = 0,
            size = 0,
            totalElements = 0L,
            totalPages = 0,
            hasNext = false,
            semInventarioAtivo = true
        )

        /**
         * Variante explícita da fábrica vazia que permite construir
         * respostas vazias com o flag `semInventarioAtivo` controlado.
         *
         * Útil para diferenciar o cenário "sem inventário ativo"
         * (`semInventarioAtivo=true`) do cenário "nenhum resultado com
         * inventário ativo presente" (`semInventarioAtivo=false`).
         */
        fun <T> empty(semInventarioAtivo: Boolean): PagedResponseDto<T> = PagedResponseDto(
            content = emptyList(),
            page = 0,
            size = 0,
            totalElements = 0L,
            totalPages = 0,
            hasNext = false,
            semInventarioAtivo = semInventarioAtivo
        )
    }
}
